### PR TITLE
Refactor resource listers

### PR DESCRIPTION
**Pull request by svenwltr**
Some heavy refactoring in preparation of #54. Also I want to get rid of the weird reflections.

Since this is a heavy change, here some notes:

This changes the interface for registering resources. Previously it was necessary to update the `GetListers` function in `resources/listers.go`. Now every resource registers itself inside of an `init` function which is defined in `resources/interface.go` (eg `register("IAMUser", ListIAMUsers)`). Since it has to provide a name for registering it isn't required anymore to get the resource type via reflections.

Also we are getting rid of the `FooNuke` types in `resources/types.go`, which just carried the session and don't really give a benefit. Now, the `ResourceLister` interface changed so it takes a session as a command. Therefore the intermediate types aren't required anymore.

### CODE REVIEW

**Comment by svenwltr** [comment]
@rebuy-de/prp-aws-nuke Please review. Please do it carefully, since this is a big change.

File: cmd/queue.go [hunk 1]
```diff
@@ -15,32 +15,34 @@ const (
 
 type Item struct {
```
Comment (bjoernhaeuser) [inline cmd/queue.go]: Should we start adding Godoc?

File: resources/listers.go [hunk 2]
```diff
@@ -1,112 +0,0 @@
-package resources
```
Comment (stephanlindauer) [inline resources/listers.go]: :heart:

File: resources/interface.go [hunk 3]
```diff
@@ -11,3 +15,13 @@ type Filter interface {
 	Resource
 	Filter() error
 }
+
+var resourceListers = make(ResourceListers)
+
+func register(name string, lister ResourceLister) {
+	resourceListers[name] = lister
```
Comment (bjoernhaeuser) [inline resources/interface.go]: would prevent adding the same twice here

File: resources/kms-aliases.go [hunk 4]
```diff
@@ -12,16 +13,22 @@ type KMSAlias struct {
 	name string
 }
 
-func (n *KMSNuke) ListAliases() ([]Resource, error) {
-	resp, err := n.Service.ListAliases(nil)
+func init() {
+	register("KMSAliase", ListKMSAliases)
```
Comment (bjoernhaeuser) [inline resources/kms-aliases.go]: i dont think that `Aliase` is incorrect :)

File: resources/interface.go (hunk 3 above)
Comment (svenwltr) [reply resources/interface.go]: done

File: resources/kms-aliases.go (hunk 4 above)
Comment (svenwltr) [reply resources/kms-aliases.go]: fixed

File: cmd/queue.go (hunk 1 above)
Comment (svenwltr) [reply cmd/queue.go]: I added a very bit of GoDoc

File: resources/ec2-eip.go [hunk 5]
```diff
@@ -1,24 +1,33 @@
...
 
 type EC2Address struct {
 	svc *ec2.EC2
 	id  string
 	ip  string
 }
 
-func (n *EC2Nuke) ListAddresses() ([]Resource, error) {
+func init() {
+	register("EC2Addresse", ListEC2Addresses)
```
Comment (bethge) [inline resources/ec2-eip.go]: ping

File: resources/ecr-repository.go [hunk 6]
```diff
@@ -12,7 +13,13 @@ type ECRRepository struct {
 	name *string
 }
 
-func (n *ECRNuke) ListRepos() ([]Resource, error) {
+func init() {
+	register("ECRRepo", ListECRRepos)
```
Comment (bethge) [inline resources/ecr-repository.go]: ping: `ECRRepository`

File: resources/elb-elb.go [hunk 7]
```diff
@@ -1,22 +1,31 @@
...
 
 type ELB struct {
 	svc  *elb.ELB
 	name *string
 }
 
-func (n *ElbNuke) ListELBs() ([]Resource, error) {
-	resp, err := n.Service.DescribeLoadBalancers(nil)
+func init() {
+	register("ElbELB", ListElbELBs)
```
Comment (bethge) [inline resources/elb-elb.go]: `ELB`?

File: cmd/scan.go [hunk 8]
```diff
@@ -8,14 +8,12 @@ import (
...
-func Scan(sess *session.Session) <-chan *Item {
+func Scan(region Region) <-chan *Item {
 	items := make(chan *Item, 100)
 
 	go func() {
-		listers := resources.GetListers(sess)
-
-		for _, lister := range listers {
-			r, err := safeLister(lister)
+		for category, lister := range resources.GetListers() {
```
Comment (bethge) [inline cmd/scan.go]: This is by now the only place we use the variable name `category`. It is actually resource type name, right? Would love to get rid of `category` once and for all.

File: resources/elbv2-alb.go [hunk 9]
```diff
@@ -1,23 +1,32 @@
...
 type ELBv2 struct {
 	svc  *elbv2.ELBV2
 	name *string
 	arn  *string
 }
 
-func (n *Elbv2Nuke) ListELBs() ([]Resource, error) {
-	resp, err := n.Service.DescribeLoadBalancers(nil)
+func init() {
+	register("Elbv2ELB", ListElbv2ELBs)
```
Comment (bethge) [inline resources/elbv2-alb.go]: `ELBv2`?

File: resources/elbv2-targetgroup.go [hunk 10]
```diff
@@ -1,23 +1,32 @@
...
 type ELBv2TargetGroup struct {
 	svc  *elbv2.ELBV2
 	name *string
 	arn  *string
 }
 
-func (n *Elbv2Nuke) ListTargetGroups() ([]Resource, error) {
-	resp, err := n.Service.DescribeTargetGroups(nil)
+func init() {
+	register("Elbv2TargetGroup", ListElbv2TargetGroups)
```
Comment (bethge) [inline resources/elbv2-targetgroup.go]: `ELBv2TargetGroup`?

File: resources/elbv2-targetgroup.go [hunk 11]
```diff
@@ -1,23 +1,32 @@
...
+}
+
+func ListElbv2TargetGroups(sess *session.Session) ([]Resource, error) {
```
Comment (bethge) [inline resources/elbv2-targetgroup.go]: `ListELBv2TargetGroups`

File: resources/iam-policies.go [hunk 12]
```diff
@@ -10,8 +11,14 @@ type IAMPolicy struct {
 	arn string
 }
 
-func (n *IAMNuke) ListPolicies() ([]Resource, error) {
-	resp, err := n.Service.ListPolicies(&iam.ListPoliciesInput{
+func init() {
+	register("IAMPolicie", ListIAMPolicies)
```
Comment (bethge) [inline resources/iam-policies.go]: `IAMPolicy`

File: resources/kms-aliases.go [hunk 13]
```diff
@@ -12,16 +13,22 @@ type KMSAlias struct {
 	name string
 }
 
-func (n *KMSNuke) ListAliases() ([]Resource, error) {
-	resp, err := n.Service.ListAliases(nil)
+func init() {
+	register("KMSAliases", ListKMSAliases)
```
Comment (bethge) [inline resources/kms-aliases.go]: `KMSAlias`

File: resources/rds-subnets.go [hunk 14]
```diff
@@ -10,16 +11,22 @@ type RDSDBSubnetGroup struct {
 	name *string
 }
 
-func (n *RDSNuke) ListSubnetGroups() ([]Resource, error) {
+func init() {
+	register("RDSSubnetGroup", ListRDSSubnetGroups)
```
Comment (bethge) [inline resources/rds-subnets.go]: `RDSDBSubnetGroup`

File: resources/route53-resource-records.go [hunk 15]
```diff
@@ -14,17 +15,23 @@ type Route53ResourceRecordSet struct {
 	changeId     *string
 }
 
-func (n *Route53Nuke) ListResourceRecords() ([]Resource, error) {
+func init() {
+	register("Route53ResourceRecord", ListRoute53ResourceRecords)
```
Comment (bethge) [inline resources/route53-resource-records.go]: `Route53ResourceRecordSet`

**Comment by svenwltr** [comment]
I added checked the resource names against version `1.3.0`. For that I added a new command `aws-nuke resource-types`. Since that command doesn't exist on `1.3.0`, I did some magic: `ack -oh '^type (\S+) struct' | grep -v Nuke | vi -`.

File: cmd/root.go [hunk 16]
```diff
@@ -70,6 +74,28 @@ func NewRootCommand() *cobra.Command {
...
 	command.AddCommand(NewVersionCommand())
+	command.AddCommand(NewResourceTypesCommand())
 
 	return command
 }
+
+func NewResourceTypesCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "resource-types",
+		Short: "lists all availible resource types",
```
Comment (bethge) [inline cmd/root.go]: available

File: cmd/root.go [hunk 17]
```diff
@@ -70,6 +74,28 @@ func NewRootCommand() *cobra.Command {
...
+		Run: func(cmd *cobra.Command, args []string) {
+			types := []string{}
+			for resourceType, _ := range resources.GetListers() {
+				types = append(types, resourceType)
+			}
+
+			sort.Sort(sort.StringSlice(types))
```
Comment (bethge) [inline cmd/root.go]: Would this work: https://golang.org/pkg/sort/#Strings ?

**Pull request merged**